client/main/infobox.js

```diff
@@ -28,39 +29,151 @@ class InfoBox extends HTMLElement {
         this._remote = this._body.querySelector('.remote');
-        this._heading = this._body.querySelector('h1');
+        this._heading = this._body.querySelector('.heading .title');
         this._content = this._body.querySelector('.content');
-        this._button = this._body.querySelector('button');
+        this._button = this._body.querySelector('.button-box button');
+        this._close = this._body.querySelector('.close');
+
+        this._close.addEventListener('click', () => this.hide());
+        this._host.addEventListener('click', (event) => this._backgroundClicked(event));
+        this._host.addEventListener('keydown', (event) => this._onKeyDown(event));
 
         this._button.addEventListener('click', () => this.clicked());
         window.addEventListener('message', (event) => this.onMessage(event));
     }
 
-    setup(info) {
-        this._host.setAttribute('title', info.title || '');
-        this._host.setAttribute('message', info.message || '');
-        this._host.setAttribute('message-src', info['message-src'] || '');
-        this._host.setAttribute('status', info.status || '');
+    setup(info, params) {
+        if (params === undefined)
+            params = { };
+
+        if (this._displayInfo)
+            this.hide();
+
+        this._visible = true;
+        this._processEnterKey = false;
+
+        this._displayInfo = info;
+        if (info['message-src']) {
+            if (params.cancelable === undefined)
+                params.cancelable = false;
+
+            this._local.classList.remove('external');
+            this._body.style.height = `300px`;
+            this._body.style.width = `300px`;
+            this._host.setAttribute('message-src', info['message-src'] || '');
+            this._host.style.display = null;
+            keyboardJS.pause();
+        }
+        else {
+            if (params.cancelable === undefined)
+                params.cancelable = true;
+
+            this._local.classList.add('external');
+            let show = true;
+            if (status.message || info.title) {
+                this._host.setAttribute('title', info.title || '');
+                this._host.setAttribute('message', info.message || '');
+                this._host.setAttribute('status', info.status || '');
+                this._processEnterKey = true;
+            }
+            else if (this._isElement(info)) {
+                this.appendChild(info);
+            }
+            else
+                show = false;
+
+            if (show) {
+                keyboardJS.pause();
+                this._host.style.display = null;
+                setTimeout(() => {
+                    this._body.style.opacity = 1;
+                    this._host.style.opacity = 1;
+                }, 10);
+            }
+        }
+
+        this._cancelable = params.cancelable;
+
+        if (this._cancelable) {
+            this._close.style.display = 'block';
```

Comment:
should this be `this._close.style.display = null`? it looks like you're wanting to 'unset' it.